### ISSUE TITLE
Why did PR #21 disabled dependency caching on Cobalt

### ISSUE THREAD

**Issue by jshannon63** [kocsismate/php-di-container-benchmarks#24]
So, I am not sure why it was determined appropriate to disable dependency caching on the cobalt container? Avoiding unnecessary and repeated calls to ReflectionClass seems an obvious choice for performance improvement. Does this somehow invalidate the testing? In a benchmark setting, having one participants submission "tuned" by another should maybe prompt an invitation to discuss.

**Comment by kocsismate** [comment]
Did I get it right that the cached mode really means that if you retrieve a class, its dependencies are stored for later use and on the subsequent calls to the same class, the container instantiates a new object with the cached dependencies? If it isn't the case, then I apologize, and we can re-add the cached mode.

Otherwise It was appropriate to disable caching. The reason is that the cached mode is neither prototype, nor singleton mode in fact, and containers in Test Suites 3 and 4 are expected to return completely new classes with all their dependencies newly instantiated.

I am looking forward to your answer!

**Comment by jshannon63** [comment]
The only thing we cache is the reflection object ($reflect) and the dependency classnames ($dependencies). So yes, we do create and return new instances in each case.  The dependencies are created new by $reflect->newInstanceArgs($dependencies) since we are only passing in the stored dependency classnames (from getParameters) as strings. The only thing we save ourself in cached mode is repeated call to \ReflectionClass. 

This method should be not unlike working with a compiled container. Only, we just do it dynamically, at first peek, with each class.

I hope I have done ok explaining and this all makes sense. Maybe I am missing something... could be. Let me know. Of course I want it to be a fair benchmark, so If something I need to consider, please let me know.

BTW, I am surprised this has not been done before. Or, if it has I am unaware.

**Comment by kocsismate** [comment]
I would never thought that you wanted to cheat in the benchmark :)

As I am not really familiar with `ReflectionClass`, I am using `var_dump` here to debug the code:

```php
if ($this->bindings[$id]['cached'] && $this->mode == 'cached') {
            var_dump($this->bindings[$id]['reflect']);
            var_dump($this->bindings[$id]['dependencies']);
            exit;
            return $this->bindings[$id]['reflect']->newInstanceArgs($this->bindings[$id]['dependencies']);
        }
```

And when I call `$container->get(Class10::class)` the second time, I get this:

```
object(ReflectionClass)#29 (1) {
  ["name"]=>
  string(37) "DiContainerBenchmarks\Fixture\Class10"
}
array(1) {
  [0]=>
  object(DiContainerBenchmarks\Fixture\Class9)#41 (0) {
  }
}
```

It looks for me that a `Class9` instance is cached, but I am eager to hear about your investigation!

**Comment by jshannon63** [comment]
I edited my second response because I am looking and finding some issues with my implementation of cached mode. Definitely would never think about doing anything the wrong way. Just an oversight from what I can see.

I will get to the bottom of it and report back to you what I find in a new issue when ready. Sorry to trouble you for your time.

**Comment by kocsismate** [comment]
No need to worry! Probably I used the wrong expression in my last comment, but I wanted to say that I didn't think that you want to manipulate the results or anything. These oversights happen all the time. This very benchmark also had serious issues which made the results unreliable when I first published it on reddit. But now, it is much better. :) So take your time, and fix the problems. I'll be happy to update the tests when you come up with a new solution for the caching!
